stdlib/string.gr

```diff
@@ -1297,3 +1297,107 @@ export let decode = (bytes: Bytes, encoding: Encoding) => {
 export let decodeKeepBom = (bytes: Bytes, encoding: Encoding) => {
   decodeHelp(bytes, encoding, false)
 }
+
+/**
+ * Iterates over Unicode code points in a string.
+ * 
+ * @param f: The iterator function
+ * @param s: The string to iterate
+ */
+@disableGC
+export let forEachCodePoint = (f: (Number) -> Void, s: String) => {
+  let (>>>) = WasmI32.shrU
+  let (-) = WasmI32.sub
+  let (&) = WasmI32.and
+  let (<) = WasmI32.ltU
+  let (<=) = WasmI32.leU
+  let (==) = WasmI32.eq
+  let (+) = WasmI32.add
+
+  let s = WasmI32.fromGrain(s)
+
+  let byteSize = WasmI32.load(s, 4n)
+
+  let mut ptr = s + 8n
+  let end = ptr + byteSize
+  
+  let mut idx = 0n
+  while (ptr < end) {
+    let byte = WasmI32.load8U(ptr, 0n)
+    let codePointByteCount = if ((byte & 0x80n) == 0x00n) {
+      1n
+    } else if ((byte & 0xF0n) == 0xF0n) {
+      4n
+    } else if ((byte & 0xE0n) == 0xE0n) {
+      3n
+    } else {
+      2n
+    }
+
+    // Note that even if up to 4 bytes are needed to represent Unicode
+    // codepoints, this doesn't mean 32 bits. The highest allowed code point is
+    // 0x10FFFF and it should not change in future versions of Unicode. This
+    // means no more than 21 bits are necessary to represent a code point and
+    // thus we can use Grain's "simple" numbers that hold up to 31 bits and
+    // avoid heap allocations. `getCodePoint` will throw
+    // MalformedUnicode exception for values exceeding this limit.
+    let codePoint = getCodePoint(ptr)
+    f(tagSimpleNumber(codePoint))
+
+    ptr += codePointByteCount
+    idx += 1n
+  }
+}
+
+/**
+ * Iterates over Unicode code points in a string. This is the same as
+ * `forEachCodePoint`, but provides the code point's index as the
+ * second argument to the iterator function.
```

Comment:
```suggestion
 * `forEachCodePoint`, but provides the code point's index in the string
 * as the second argument to the iterator function.
```

Reply:
@ospencer I think I would re-word it to this, since for some reason my first read of this made me think "code point's index" was the code point (in the sense of it's index in Unicode or something). I knew what you meant, but I think this makes it more explicit